Improve support for rpy

gabrielebndn opened the pull request:
This PR improves support for RPY by adding the following three functions:
- `rpyToJac`: Computes the Jacobian of the Roll-Pitch-Yaw conversion given the RPY angles, in reference frame either `LOCAL` or `WORLD`, defauting to `LOCAL`, plus `LOCAL_WORLD_ALIGNED` which is equivalent to `WORLD`
- `rpyToJacInv`: the inverse of such Jacobian, in given reference frame. I thought it might be faster and more stable next to singularities with respect to directly computing the inverse, but in practice I did not find a huge difference
- `rpyToJacDerivative`, time derivative of such Jacobian, given the RPY angles and their time derivative, in given reference frame.

All functions have python bindings.

- `rpyToJac` is tested against identity at RPY=0, for relationship between different frames, against the analytical formulas, and against finite differences
- `rpyToJacInv` is tested against the inverse of `rpyToJac`
- `rpyToJac` is tested against zero at dRPY/dt=0, for relationship between different frames, and against finite differences
- Python bindings are extensively tested too

jcarpent commented on src/math/rpy.hpp:
```diff
@@ -63,34 +53,70 @@ namespace pinocchio
+    ///
+    template<typename Vector3Like0, typename Vector3Like1>
+    inline Eigen::Matrix<typename Vector3Like0::Scalar,3,3,PINOCCHIO_EIGEN_PLAIN_TYPE(Vector3Like0)::Options>
+    rpyToJacDerivative(const Eigen::MatrixBase<Vector3Like0> & rpy, const Eigen::MatrixBase<Vector3Like1> & rpydot, const ReferenceFrame rf=LOCAL);
```
Adding time should ease the reader.

gabrielebndn replied:
Got it, computeRpyJacobianDerivative

jcarpent reviewed (changes requested):
Hi @gabrielebndn,

Thanks for your contribution. I made some remarks which should help to enhance this PR.
By the way, I just hope you are not abusing from Euler representation for 3d rotations: it is a very bad habit. Pinocchio is really trying to avoid this.

Best,

Justin

jcarpent commented on src/math/rpy.hpp:
```diff
@@ -7,6 +7,7 @@
 #include "pinocchio/math/comparison-operators.hpp"
+#include "pinocchio/multibody/fwd.hpp"
 
 #include <Eigen/Geometry>
```
This should be moved to the implementation file.

gabrielebndn replied:
Done

gabrielebndn commented:
> By the way, I just hope you are not abusing from Euler representation for 3d rotations: it is a very bad habit. Pinocchio is really trying to avoid this.

I do agree with you. Trying as hard as I can :slightly_smiling_face:

gabrielebndn commented on src/math/rpy.hpp:
```diff
@@ -42,12 +36,8 @@ namespace pinocchio
-      PINOCCHIO_ASSERT_MATRIX_SPECIFIC_SIZE(Vector3Like, rpy, 3, 1);
-      return rpyToMatrix(rpy[0], rpy[1], rpy[2]);
-    }
+    inline Eigen::Matrix<typename Vector3Like::Scalar,3,3,PINOCCHIO_EIGEN_PLAIN_TYPE(Vector3Like)::Options>
```
Oh, hadn't thought about it. Well, _all_ of Pinocchio's functions are template functions, do I guess it can be removed everywhere :) Anyhow, I'll remove it here

jcarpent replied:
Yes, you're right. I plan to remove them in the future big release (aka Pinocchio 3).

gabrielebndn commented on src/math/rpy.hpp:
```diff
@@ -63,34 +53,70 @@ namespace pinocchio
+    ///
+    template<typename Vector3Like>
+    inline Eigen::Matrix<typename Vector3Like::Scalar,3,3,PINOCCHIO_EIGEN_PLAIN_TYPE(Vector3Like)::Options>
+    rpyToJac(const Eigen::MatrixBase<Vector3Like> & rpy, const ReferenceFrame rf=LOCAL);
```
I was hoping to keep it short to type. But whatever, I'll change the three names to:
- computeRpyJacobian
- computeRpyJacobianInverse
- computeRpyJacobianDerivative

gabrielebndn commented on src/math/rpy.hpp:
```diff
@@ -63,34 +53,70 @@ namespace pinocchio
+    ///
+    template<typename Vector3Like>
+    inline Eigen::Matrix<typename Vector3Like::Scalar,3,3,PINOCCHIO_EIGEN_PLAIN_TYPE(Vector3Like)::Options>
+    rpyToJacInv(const Eigen::MatrixBase<Vector3Like> & rpy, const ReferenceFrame rf=LOCAL);
```
Yes, computeRpyJacobianInverse

gabrielebndn commented on src/math/rpy.hxx:
```diff
@@ -0,0 +1,187 @@
+    rpyToJac(const Eigen::MatrixBase<Vector3Like> & rpy, const ReferenceFrame rf)
+    {
+      typedef typename Vector3Like::Scalar Scalar;
+      Eigen::Matrix<Scalar,3,3> J;
```
Oh, right, fixed

gabrielebndn commented:
@jcarpent all done

jcarpent reviewed (approved):
Thanks a lot @gabrielebndn for addressing all these comments in every effective and clean way. Good job, as usual.

gabrielebndn commented:
> Thanks a lot @gabrielebndn for addressing all these comments in every effective and clean way. Good job, as usual.

You are welcome :)

By the way, codacy is complaining, but it's nonsense. When I use `random()` it says "Standard pseudo-random generators are not suitable for security/cryptographic purposes", but I am definitely not using it for security/cryptographic purposes and it is good enough for me! You can safely ignore it
